sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb

```diff
@@ -0,0 +1,130 @@
+        #
+        # max_export_attempts attempts are made to export each batch, while
+        # export fails with {FAILED_RETRYABLE}, backing off linearly in 100ms
+        # increments.
+        class BatchSpanProcessor
+          SCHEDULE_DELAY_MILLIS = 5
+          MAX_QUEUE_SIZE = 2048
+          MAX_EXPORT_BATCH_SIZE = 512
+          MAX_EXPORT_ATTEMPTS = 5
+          private_constant(:SCHEDULE_DELAY_MILLIS, :MAX_QUEUE_SIZE, :MAX_EXPORT_BATCH_SIZE, :MAX_EXPORT_ATTEMPTS)
+
+          def initialize(exporter:,
+                         schedule_delay_millis: SCHEDULE_DELAY_MILLIS,
+                         max_queue_size: MAX_QUEUE_SIZE,
+                         max_export_batch_size: MAX_EXPORT_BATCH_SIZE,
+                         max_export_attempts: MAX_EXPORT_ATTEMPTS)
+            raise ArgumentError if max_export_batch_size > max_queue_size
+
+            @exporter = exporter
+            @mutex = Mutex.new
+            @condition = ConditionVariable.new
+            @keep_running = true
+            @delay_seconds = schedule_delay_millis / 1000.0
+            @max_queue_size = max_queue_size
+            @batch_size = max_export_batch_size
+            @export_attempts = max_export_attempts
+            @spans = []
+            @thread = Thread.new { work }
+          end
+
+          # does nothing for this processor
+          def on_start(span)
+            # noop
+          end
+
+          # adds a span to the batcher, threadsafe may block on lock
+          def on_end(span)
+            return unless span.recording_events?
+
+            lock do
+              n = spans.size + 1 - max_queue_size
+              spans.shift(n) if n.positive?
+              spans << span
+              @condition.signal if spans.size > max_queue_size / 2
+            end
+          end
+
+          # shuts the consumer thread down and flushes the current accumulated buffer
+          # will block until the thread is finished
+          def shutdown
+            lock do
+              @keep_running = false
+              @condition.signal
+            end
+
+            @thread.join
+            flush
+            @exporter.shutdown
+          end
+
+          private
+
+          attr_reader :spans, :max_queue_size, :batch_size
+
+          def work
+            loop do
+              batch = lock do
+                @condition.wait(@mutex, @delay_seconds) if spans.size < batch_size && @keep_running
+                @condition.wait(@mutex, @delay_seconds) while spans.empty? && @keep_running
+                return unless @keep_running
+
+                fetch_batch
+              end
+
+              export_batch(batch) # TODO: log or count errors
+            end
+          end
+
+          def export_batch(batch)
+            @export_attempts.times do |retries|
+              result_code = @exporter.export(batch)
+              return result_code unless result_code == FAILED_RETRYABLE
+
+              sleep(0.1 * retries)
```

Comment:
Is the wait between retries something that should be configurable? Should it be a parameter to the constructor like `wait_time_millis`and friends?

Reply:
The spec says nothing about limiting retries, or how backoff should be handled. I'm not a huge fan of adding too many knobs here. Is there a way we could punt all of the backoff & retry logic to the user, with a reasonable default? I.e. allow them to provide a proc.

Reply:
I think we can punt for now and consider allowing a user specified proc in the future.